src/python/pants/backend/js/target_types.py

```diff
@@ -0,0 +1,54 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+from pants.engine.target import (
+    COMMON_TARGET_FIELDS,
+    Dependencies,
+    MultipleSourcesField,
+    SingleSourceField,
+    Target,
+    TargetFilesGenerator,
+)
+
+JS_FILE_EXTENSIONS = (".js",)
+
+
+class JSSourceField(SingleSourceField):
+    expected_file_extensions = JS_FILE_EXTENSIONS
+
+
+class JSGeneratorSourcesField(MultipleSourcesField):
+    expected_file_extensions = JS_FILE_EXTENSIONS
+
+
+# -----------------------------------------------------------------------------------------------
+# `js_source` and `js_sources` targets
+# -----------------------------------------------------------------------------------------------
+
+
+class JSSourceTarget(Target):
+    alias = "js_source"
+    core_fields = (
+        *COMMON_TARGET_FIELDS,
+        Dependencies,
+        JSSourceField,
+    )
+    help = "A single Javascript source file."
+
+
+class JSSourcesGeneratorSourcesField(JSGeneratorSourcesField):
+    default = tuple(f"*{ext}" for ext in JS_FILE_EXTENSIONS)
```

Comment:
We will eventually have test targets. I think a common pattern is `foo.test.js`, right? 

Because this is so experimental, it's fine to have those belong to `js_sources` -- but do note that will be a breaking change when we change the default sources to exclude test files.

Excluding eagerly isn't great either, as it means test files wouldn't have Prettier run on them as they'd be unowned.

Reply:
Yep, some default test targets will come in a later PR - when I can review a few of the common options for test naming. `.test.js` is what I use, there is also `.spec.js` and a couple more I think.